File: javascript/process.js
const _ = require('lodash');

const constants = require('./constants');
const FileHandler = require('./file');
const Template = require('./template');

class WWE {
  constructor() {
    const championships = FileHandler.readJson('championships');
    this.primaryChampionships = championships.primary;
    this.secondaryChampionships = championships.secondary;
    this.tertiaryChampionships = championships.tertiary;
    this.tagTeamChampionships = championships.tag_team;
    this.womenChampionships = championships.women;
    this.womenTagTeamChampionships = championships.women_tag_team;
    this.singleChampionships = []
        .concat(this.primaryChampionships)
        .concat(this.secondaryChampionships)
        .concat(this.tertiaryChampionships);
    this.maleChampionships = this.singleChampionships
        .concat(this.tagTeamChampionships);
    this.femaleChampionships = []
        .concat(this.womenChampionships)
        .concat(this.womenTagTeamChampionships);
    this.maleWrestlers = {};
    this.femaleWrestlers = {};
  }

  calculatePoints() {
    let p, s, t, tt;
    Object.keys(this.maleWrestlers).forEach((key) => {
      p = s = t = tt = 0;
      const wrestler = this.maleWrestlers[key];
      this.primaryChampionships.forEach(c => p+=wrestler[c]);
      this.secondaryChampionships.forEach(c => s+=wrestler[c]);
      this.tertiaryChampionships.forEach(c => t+=wrestler[c]);
      this.tagTeamChampionships.forEach(c => tt+=wrestler[c]);
      wrestler['primary'] = p;
      wrestler['secondary'] = s;
      wrestler['tertiary'] = t;
      wrestler['tag_team'] = tt;
      wrestler['points'] = 3*p + 2*s + t + tt;
      wrestler['total'] = p + s + t + tt;
    });
    Object.keys(this.femaleWrestlers).forEach((key) => {
      p = tt = 0;
      const wrestler = this.femaleWrestlers[key];
      this.womenChampionships.forEach(c => p+=wrestler[c]);
      this.womenTagTeamChampionships.forEach(c => tt+=wrestler[c]);
      wrestler['primary'] = p;
      wrestler['tag_team'] = tt;
      wrestler['points'] = 2*p + tt;
      wrestler['total'] = p + tt;
    });
  }

  addChampionshipToWrestler(wrestlers, wrestler, template, championship) {
    if (!wrestlers.hasOwnProperty(wrestler)) {
      wrestlers[wrestler] = Object.assign({}, template);
    }
    wrestlers[wrestler][championship]++;
  }

  evaluateSingleChampionship(row, wrestlers, template, championships) {
    championships.forEach((championship) => {
      const wrestler = row[championship];
      if(row[championship]) {
        this.addChampionshipToWrestler(wrestlers, wrestler, template, championship);
      }
    });
  }

  evaluateTagTeamChampionship(row, wrestlers, template, championships) {
    championships.forEach((championship) => {
      if(row[championship]) {
        const x = row[championship].split('\/');
        const wrestler1 = x[0];
        const wrestler2 = x[1];
        this.addChampionshipToWrestler(wrestlers, wrestler1, template, championship);
        this.addChampionshipToWrestler(wrestlers, wrestler2, template, championship);
      }
    });
  }

  evaluate(events) {
    const maleTemplate = Template.male(this.maleChampionships);
    const femaleTemplate = Template.female(this.femaleChampionships);
    Object.freeze(maleTemplate);
    Object.freeze(femaleTemplate);
    events.forEach((row) => {
      console.log(`Evaluating event: ${row.Event}`);
      this.evaluateSingleChampionship(row, this.maleWrestlers, maleTemplate, this.singleChampionships);
      this.evaluateSingleChampionship(row, this.femaleWrestlers, femaleTemplate, this.womenChampionships);
      this.evaluateTagTeamChampionship(row, this.maleWrestlers, maleTemplate, this.tagTeamChampionships);
      this.evaluateTagTeamChampionship(row, this.femaleWrestlers, femaleTemplate, this.womenTagTeamChampionships);
    });
  }

  getSortedArrFromObj(wrestlersObj, sortBy) {
    const wrestlerArr = [];
    Object.keys(wrestlersObj).forEach((key) => {
      const obj = { wrestler: key };
      Object.assign(obj, wrestlersObj[key]);
      wrestlerArr.push(obj);
    });

    const order = Array(sortBy.length).fill('desc');
    return _.orderBy(wrestlerArr, sortBy, order);
  }

  async process() {
    const events = await FileHandler.readCsv('events');
    this.evaluate(events);
    this.calculatePoints();

    const maleWrestlersArr = this.getSortedArrFromObj(this.maleWrestlers, constants.maleFields);
    const femaleWrestlersArr = this.getSortedArrFromObj(this.femaleWrestlers, constants.femaleFields);

    FileHandler.writeCsv('male', maleWrestlersArr);
    FileHandler.writeCsv('female', femaleWrestlersArr);
    FileHandler.writeJson('male', maleWrestlersArr);
    FileHandler.writeJson('female', femaleWrestlersArr);
  }
}

module.exports = WWE;

(function () {
  const wwe = new WWE();
  wwe.process().then(() => console.log('done'));
})();

File: python/process.py
from operator import itemgetter as i
from functools import cmp_to_key
from python.constants import Constants
from python.file import FileHandler
from python.template import Template


class WWE:
    def __init__(self):
        championships = FileHandler.read_json('championships')
        self.primary_championships = championships['primary']
        self.secondary_championships = championships['secondary']
        self.tertiary_championships = championships['tertiary']
        self.tag_team_championships = championships['tag_team']
        self.women_championships = championships['women']
        self.women_tag_team_championships = championships['women_tag_team']
        self.single_championships = \
            self.primary_championships + \
            self.secondary_championships + \
            self.tertiary_championships
        self.male_championships = self.single_championships + self.tag_team_championships
        self.female_championships = self.women_championships + self.women_tag_team_championships
        self.male_wrestlers = {}
        self.female_wrestlers = {}

    def calculate_points(self):
        for wrestler, data in self.male_wrestlers.items():
            p = sum([data[c] for c in self.primary_championships])
            s = sum([data[c] for c in self.secondary_championships])
            t = sum([data[c] for c in self.tertiary_championships])
            tt = sum([data[c] for c in self.tag_team_championships])
            data['primary'] = p
            data['secondary'] = s
            data['tertiary'] = t
            data['tag_team'] = tt
            data['points'] = 3 * p + 2 * s + t + tt
            data['total'] = p + s + t + tt

        for wrestler, data in self.female_wrestlers.items():
            p = sum([data[c] for c in self.women_championships])
            tt = sum([data[c] for c in self.women_tag_team_championships])
            data['primary'] = p
            data['tag_team'] = tt
            data['points'] = 2 * p + tt
            data['total'] = p + tt

    @staticmethod
    def add_championship_to_wrestler(wrestlers, wrestler, template, championship):
        if wrestler not in wrestlers:
            wrestlers[wrestler] = dict(template)
        wrestlers[wrestler][championship] += 1

    def evaluate_single_championship(self, event, wrestlers, template, championships):
        for championship in championships:
            if championship in event and event[championship] != '':
                wrestler = event[championship]
                self.add_championship_to_wrestler(wrestlers, wrestler, template, championship)

    def evaluate_tag_team_championship(self, event, wrestlers, template, championships):
        for championship in championships:
            if championship in event and event[championship] != '':
                x = event[championship].split('/')
                wrestler1, wrestler2 = x[0], x[1]
                self.add_championship_to_wrestler(wrestlers, wrestler1, template, championship)
                self.add_championship_to_wrestler(wrestlers, wrestler2, template, championship)

    def evaluate(self):
        events = FileHandler.read_csv('events')
        male_template = Template.male(self.male_championships)
        female_template = Template.female(self.female_championships)
        for row in events:
            print('Evaluating event: ', row['Event'])
            self.evaluate_single_championship(row, self.male_wrestlers, male_template, self.single_championships)
            self.evaluate_single_championship(row, self.female_wrestlers, female_template, self.women_championships)
            self.evaluate_tag_team_championship(row, self.male_wrestlers, male_template, self.tag_team_championships)
            self.evaluate_tag_team_championship(row, self.female_wrestlers, female_template,
                                                self.women_tag_team_championships)

    @staticmethod
    def get_sorted_arr_from_obj(wrestler_obj: dict, columns: list) -> list:
        wrestler_arr = []
        for wrestler, data in wrestler_obj.items():
            obj = {'wrestler': wrestler}
            for k, v in data.items():
                obj[k] = v
            wrestler_arr.append(obj)

        comparers = [
            ((i(col[1:].strip()), -1) if col.startswith('-') else (i(col.strip()), 1))
            for col in columns
        ]

        def cmp(a, b):
            return (a > b) - (a < b)

        def comparer(left, right):
            comparer_iter = (
                cmp(fn(left), fn(right)) * mult
                for fn, mult in comparers
            )
            return next((result for result in comparer_iter if result), 0)
        return sorted(wrestler_arr, key=cmp_to_key(comparer))

    def process(self):
        self.evaluate()
        self.calculate_points()
        male_wrestlers_arr = self.get_sorted_arr_from_obj(self.male_wrestlers, Constants.sort_male_fields)
        female_wrestlers_arr = self.get_sorted_arr_from_obj(self.female_wrestlers, Constants.sort_female_fields)

        FileHandler.write_csv('male', male_wrestlers_arr)
        FileHandler.write_csv('female', female_wrestlers_arr)


if __name__ == "__main__":
    wwe = WWE()
    wwe.process()
